sklearn/preprocessing/imputation.py

```diff
@@ -372,4 +409,15 @@ def transform(self, X):
 
             X[coordinates] = values
 
+            if self.add_missing_indicator:
+                features_with_missing_values = np.where(np.any
+                                                        (mask, axis=0))[0]
+                imputed_mask = mask[:, features_with_missing_values]
+                X = np.hstack((X, imputed_mask))
+                if self.axis == 0:
+                    self.imputed_features_ = valid_idx
+                    [features_with_missing_values]
```

Comment:
One option is just to make separate methods for handling the sparse and dense cases, hence avoiding deep indentation.